File: 2024/02/aoc_24_2_2.py
# AoC 2024 - Day 2 - Task 2
# https://erik-skopp.de/AdventofCode/2024/2/
# https://github.com/eskopp/AdventOfCode/blob/main/2024/01/aoc24_2_2.py
#
# Day 2 - Red-Nosed Reports
# result: 


from os import chdir
from os.path import join

# Change the working directory
chdir(join("2024", "02"))

# Load the file and convert the data into integers
data = []
with open("example.in", "r") as file:
    for line in file:
        # Remove line breaks and convert the line into a list of integers
        data.append(list(map(int, line.strip().split(" "))))

# Function to check if a list is safe
def is_safe(levels):
    # Check if the list is sorted either ascending or descending
    if levels == sorted(levels) or levels == sorted(levels, reverse=True):
        # Check the differences between adjacent numbers
        differences = [abs(levels[i] - levels[i+1]) for i in range(len(levels)-1)]
        return all(1 <= diff <= 3 for diff in differences)
    return False

# Count the safe reports
safe_count = sum(1 for report in data if is_safe(report))

# Print the result
print(f"Number of safe reports: {safe_count}")

File: 2024/02/aoc_24_2_2.m
% AoC 2024 - Day 2 - Task 2
% https:%erik-skopp.de/AdventofCode/2024/2/
% https:%github.com/eskopp/AdventOfCode/blob/main/2024/01/aoc24_2_2.m
%
% Day 2 - Red-Nosed Reports
% result: 366

% Load data from the input file
fileID = fopen('example.in', 'r');
data = textscan(fileID, '%s', 'Delimiter', '\n');
fclose(fileID);

% Convert data into a cell array of numeric arrays
data = cellfun(@(x) str2double(strsplit(x)), data{1}, 'UniformOutput', false);

% Function to check if a report is safe
is_safe = @(levels) ...
    (issorted(levels) || issorted(levels, 'descend')) && ...
    all(abs(diff(levels)) >= 1 & abs(diff(levels)) <= 3);

% Function to check if a report can be made safe by removing one level
can_be_made_safe = @(levels) any(arrayfun(@(i) ...
    is_safe([levels(1:i-1), levels(i+1:end)]), 1:length(levels)));

% Count the number of safe reports (including Problem Dampener)
safe_count = sum(cellfun(@(report) is_safe(report) || can_be_made_safe(report), data));

% Display the result
fprintf('Number of safe reports with Problem Dampener: %d\n', safe_count);
